File: solutions/python/triangle/1/triangle.py
def sides_greater_than_zero(sides: list) -> bool:
    """Determine if all sides are greater than zero"""
    for side in sides:
        if side <= 0:
            return False

    return True


def triangle_equality_rule(sides: list) -> bool:
    """Determine if the sides follow the Triangle Sides Rule"""
    if sides[0] <= sides[1] + sides[2]:
        if sides[1] <= sides[0] + sides[2]:
            if sides[2] <= sides[0] + sides[1]:
                return True

    return False


def viable_triangle(sides: list) -> bool:
    """Determine if the values in sides makes a triangle viable"""
    if sides_greater_than_zero(sides):
        if triangle_equality_rule(sides):
            return True

    return False


def equilateral(sides: list) -> bool:
    """Determine if sides result in Equilateral Triangle"""
    if viable_triangle(sides):
        if len(set(sides)) == 1:
            return True

    return False


def isosceles(sides: list) -> bool:
    """Determine if sides result in Isosceles Triangle"""
    if viable_triangle(sides):
        if not scalene(sides):
            return len(set(sides)) >= 1

    return False


def scalene(sides: list) -> bool:
    """Determine if sides result in Scalene triangle"""
    # Scalene Triangles have three unique sides
    # So if we convert the side List to a Set, we should count
    # 3 elements for a Scalene. Additionally, the length of the
    # set should be equal to the length of the list

    if viable_triangle(sides):
        return len(sides) == len(set(sides))
    else:
        return False

File: solutions/cpp/triangle/1/triangle.cpp
#include "triangle.h"
#include <stdexcept>
#include <unordered_map>

namespace triangle {

    bool isTriangle(float a, float b, float c) {

        if ((a == 0) || (b == 0) || (c == 0)) {
            return false;
        } else if ((a + b >= c) && (b + c >= a) && (a+ c >= b)) {
            return true;
        } else {
            return false;
        }
    }

    flavor kind(float a, float b, float c) {

        std::unordered_map<float, int> sides;
        sides[a]++;
        sides[b]++;
        sides[c]++;
        
        if (!isTriangle(a, b, c)){
            throw std::domain_error("Inputs do not make a Triangle");
        }

        for (const std::pair<float, int> side : sides) {
            if (side.second == 3) {
                return flavor::equilateral;
            } else if (side.second == 2) {
                return flavor::isosceles;
            }
        }
        return flavor::scalene;
    }

}  // namespace triangle
